### PR TITLE
Add a bunch of C-grid infrastructure

### PR DESCRIPTION
**Pull request by apcraig**
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    apcraig
- [X] Developer(s): 
    Add C-grid infrastructure
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    TBD
- How much do the PR code changes differ from the unmodified code? 
    - [ ] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [X] Yes
    - [ ] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [X] Yes
    - [ ] No, does the documentation need to be updated at a later time?
        - [X] Yes
        - [ ] No 
- [ ] Please provide any additional information or relevant details below:

This is a first proposal of C-grid infrastructure changes.

So far, this adds

- N and E grids, lengths, areas, masks
- Adds new grid variables to history file
- Adds grid_average_X2Y methods for averaging between all grids (i.e. t2u)
- Add gridavgchk unit tester

Still to be done

- Complete and validate X2Y methods fully
- Add grid_system namelist
- Add BC masks
- Documentation still needs to be updated.  (See https://apcraig-cice.readthedocs.io/en/cgrida/index.html)
- Testing still needs to be done. Seems to be mostly bit-for-bit so far, trying to maintain that. PGI on cheyenne seems to change answers (as usual).

### CODE REVIEW

File: cicecore/cicedynB/infrastructure/ice_grid.F90 [hunk 1]
```diff
@@ -448,6 +449,10 @@ subroutine init_grid2
 
       !-----------------------------------------------------------------
       ! T-grid cell and U-grid cell quantities
+      ! Fill halo data locally where possible to avoid missing
```
Comment (eclare108213) [inline cicecore/cicedynB/infrastructure/ice_grid.F90:452]: Is this change BFB when blocks are not eliminated?

File: cicecore/cicedynB/analysis/ice_history_shared.F90 [hunk 2]
```diff
@@ -152,32 +152,52 @@ module ice_history_shared
...
          ucstr = 'area: uarea'          , & ! vcellmeas for U cell quantities
+         ncstr = 'area: narea'          , & ! vcellmeas for N cell quantities
+         ecstr = 'area: earea'          , & ! vcellmeas for E cell quantities
          tstr2D  = 'TLON TLAT time'     , & ! vcoord for T cell quantities, 2D
          ustr2D  = 'ULON ULAT time'     , & ! vcoord for U cell quantities, 2D
+         nstr2D  = 'NLON NLAT time'     , & ! vcoord for N cell quantities, 2D
+         estr2D  = 'ELON ELAT time'     , & ! vcoord for E cell quantities, 2D
          tstr3Dz = 'TLON TLAT VGRDi time',& ! vcoord for T cell quantities, 3D
          ustr3Dz = 'ULON ULAT VGRDi time',& ! vcoord for U cell quantities, 3D
+         nstr3Dz = 'NLON NLAT VGRDi time',& ! vcoord for N cell quantities, 3D
```
Comment (eclare108213) [inline cicecore/cicedynB/analysis/ice_history_shared.F90:169]: I don't think we will need some of these options on the N and E grids, but it's okay to have them available for future use.  The B- vs C-grid choice should only impact the dynamics variables, which do not have vertical, ITD (ncat) or FSD dimensions.

File: doc/source/user_guide/ug_implementation.rst [hunk 3]
```diff
@@ -452,7 +452,9 @@ block equally.  This is useful in POP which always has work in
 each block and is written with a lot of
 array syntax requiring calculations over entire blocks (whether or not
 land is present).  This option is provided in CICE as well for 
-direct-communication compatibility with POP. The ‘latitude’ option 
+direct-communication compatibility with POP. Block that contain 100%
+land grid cells are eliminated with 'block'.  The 'blockall' option is identical
```
Comment (eclare108213) [inline doc/source/user_guide/ug_implementation.rst:509]: I don't understand the sentence that begins with 'Block'.

**Review by eclare108213 — COMMENTED** [review]
Lots of progress already!

**Review by dabail10 — APPROVED** [review]
This somehow looks familiar. Looking good.

**Comment by apcraig** [comment]
Documentation can be viewed here, https://apcraig-cice.readthedocs.io/en/cgrida/index.html

File: cicecore/cicedynB/infrastructure/ice_grid.F90 (hunk 1 above)
Comment (apcraig) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:452]: I've run the quick suite on cheyenne pgi, intel, and gnu and it is BFB so far.  I was worried about this too.  By avoiding the haloupdate for the variables, we can do a little more in terms of the averaging to/from grids for blocks where the neighbor is eliminated.  I need to test the full suite and will do that tonight.

File: cicecore/cicedynB/analysis/ice_history_shared.F90 (hunk 2 above)
Comment (apcraig) [reply cicecore/cicedynB/analysis/ice_history_shared.F90:169]: These 4d options already exists for the ULON ULAT, so that's why I added them for the N and E grids.  It could be that we can get rid of some of them at some point.

File: doc/source/user_guide/ug_implementation.rst (hunk 3 above)
Comment (apcraig) [reply doc/source/user_guide/ug_implementation.rst:509]: I fixed it, it should read "Blocks that contain 100% land grid cells are eliminated with 'block'.  The 'blockall' option is identical to 'block' except it does not do land block elimination."  Let me know if that's still not clear.

**Comment by apcraig** [comment]
I think this could be merged.  Test results are here, https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#4cd4039ab1d7a25b6a5af991edf1b67eb011a5dc.  Everything looks good.  Some of the pgi results are not bit-for-bit, but that's typical for pgi.  gnu and intel are both bit-for-bit for all tests.  The only think missing is the boundary condition masks, that could be done as a separate PR.  Lots of things here may need to be revised as we implement the C/CD grid including

- more grid variables for particular length scales that should be precomputed
- update the X2Y grid mapping to support particular weighting or masks needed
- remove/refactor the new "extra" uveln and vvele arrays that are needed for the CD grid
- update history variables that changes underlying grids in the CD solver

File: cicecore/cicedynB/analysis/ice_history.F90 [hunk 4]
```diff
@@ -538,22 +568,32 @@ subroutine init_hist (dt)
...
              "ice velocity (x)",                                  &
-             "positive is x direction on U grid", c1, c0,         &
+             "positive is x direction on u grid", c1, c0,         &
              ns1, f_uvel)
       
-         call define_hist_field(n_vvel,"vvel","m/s",ustr2D, ucstr,  &
+         call define_hist_field(n_vvel,"vvel","m/s",l_vstr2D, l_vcstr,  &
              "ice velocity (y)",                                  &
-             "positive is y direction on U grid", c1, c0,         &
+             "positive is y direction on v grid", c1, c0,         &
```
Comment (eclare108213) [inline cicecore/cicedynB/analysis/ice_history.F90]: There's not a "v" grid, is there?  I think this line and the uvel one above should remain as "on U grid".

File: cicecore/cicedynB/analysis/ice_history.F90 [hunk 5]
```diff
@@ -538,22 +568,32 @@ subroutine init_hist (dt)
...
+             ns1, f_uveln)
+      
+         call define_hist_field(n_vvele,"vvele","m/s",estr2D, ecstr,  &
+             "ice velocity (y)",                                  &
+             "positive is y direction on E grid", c1, c0,         &
+             ns1, f_vvele)
+
          call define_hist_field(n_uatm,"uatm","m/s",ustr2D, ucstr,  &
              "atm velocity (x)",                                  &
              "positive is x direction on U grid", c1, c0,         &
```
Comment (eclare108213) [inline cicecore/cicedynB/analysis/ice_history.F90]: Is this comment correct, that uatm, vatm are on the U grid?  They might actually be on the T grid here -- need to check (and if wrong, this could be fixed in a subsequent history PR).

File: cicecore/cicedynB/analysis/ice_history_shared.F90 (hunk 2 above)
Comment (eclare108213) [reply cicecore/cicedynB/analysis/ice_history_shared.F90:169]: Let's add a "clean up history" checkbox in the project board, to remind us to look back at this.  I'd suggest commenting out whatever is not immediately needed.

File: cicecore/cicedynB/dynamics/ice_dyn_shared.F90 [hunk 6]
```diff
@@ -162,6 +163,10 @@ subroutine init_dyn (dt)
          ! velocity
          uvel(i,j,iblk) = c0    ! m/s
          vvel(i,j,iblk) = c0    ! m/s
+         if (grid_system == 'CD') then ! extra velocity variables
```
Comment (eclare108213) [inline cicecore/cicedynB/dynamics/ice_dyn_shared.F90:166]: Just noting here that you're reusing uvel, vvel on E, N faces for the C-grid.  That's not clear in the history output.

File: cicecore/cicedynB/analysis/ice_history.F90 (hunk 4 above)
Comment (eclare108213) [reply cicecore/cicedynB/analysis/ice_history.F90]: Okay, now I see that you're reusing uvel, vvel for the C-grid.  So maybe to be clearer, there could be a conditional statement that sets the comment depending on which grid is in use, e.g. keep original for B-grid but for C-grid something like "positive is X direction at East face of grid cell"?

File: cicecore/cicedynB/infrastructure/ice_grid.F90 [hunk 7]
```diff
@@ -59,19 +65,29 @@ module ice_grid
          dyt    , & ! height of T-cell through the middle (m)
          dxu    , & ! width of U-cell through the middle (m)
          dyu    , & ! height of U-cell through the middle (m)
+         dxn    , & ! width of N-cell through the middle (m)
+         dyn    , & ! height of N-cell through the middle (m)
+         dxe    , & ! width of E-cell through the middle (m)
```
Comment (eclare108213) [inline cicecore/cicedynB/infrastructure/ice_grid.F90:70]: Note that dxe (i,j) = HUS (i,j).  If I remember correctly, HUS etc are available in the grid files (is that right)?  If so, are these calculations consistent with those?

File: cicecore/cicedynB/infrastructure/ice_grid.F90 [hunk 8]
```diff
@@ -1819,11 +1949,90 @@ subroutine Tlatlon
 
             ! TLAT in radians North
             TLAT(i,j,iblk) = asin(tz)
+
+! these two loops should be merged to save cos/sin calculations,
```
Comment (eclare108213) [inline cicecore/cicedynB/infrastructure/ice_grid.F90:1953]: If this routine isn't called during the timestepping, then this bit of savings won't matter much.  I don't remember why it's like this, but I do remember that this subroutine is super finicky.  Trig calculations are required based on the center of the Earth for accuracy.  I'm not sure that simple averaging is good enough - need to look at it more closely.  Might be good enough for now, though.

File: cicecore/cicedynB/infrastructure/ice_grid.F90 [hunk 9]
```diff
@@ -1872,154 +2105,381 @@ subroutine Tlatlon
...
-
-!=======================================================================
-
-! Transfer from U-cell centers to T-cell centers. Writes work into
-! another array that has ghost cells
-! NOTE: Input array is dimensioned only over physical cells.
-!
-! author: Elizabeth C. Hunke, LANL
-
-      subroutine u2tgrid_vector (work)
```
Comment (eclare108213) [inline cicecore/cicedynB/infrastructure/ice_grid.F90:1981]: The "_vector" suffix would indicate to me that this (old) routine would calculate vector averages, which it does not.  I'm guessing that it was called this because vector quantities are on the U grid (for the B-grid case), which needs the extra halo update before shifting.  (I looked at the 'blame' function and this one is on me...  sometime between 8 and 15 years ago).  This PR replaces this subroutine with "grid_average*", which is more appropriately named and should reproduce the current functionality on B grids.  As part of the C-grid development and testing, it would be good to check whether/how much doing proper vector averages would make a difference in the solution.  I suspect a lot, since the difference between B- and C-/CD-grids is all about the vector quantities.

File: doc/source/user_guide/ug_case_settings.rst [hunk 10]
```diff
@@ -239,6 +239,9 @@ grid_nml
    "``grid_file``", "string", "name of grid file to be read", "'unknown_grid_file'"
    "``grid_format``", "``bin``", "read direct access grid and kmt files", "``bin``"
    "", "``nc``", "read grid and kmt files", ""
+   "``grid_system``", "``B``", "use B grid structure with T at center and U at NE corner, "``B``"
+   "", "``C``", "use C grid structure with T at center, U at E edge, and V at N edge", ""
```
Comment (eclare108213) [inline doc/source/user_guide/ug_case_settings.rst:243]: We might need to be more specific in how we reference the grids.  In my current understanding, we will only implement the sea ice equations for the CD grid, but we will allow ocean coupling directly with either a C or CD grid.  Maybe we should use CD everywhere, since it's our main grid, and only mention C-grid in reference to coupling?

File: doc/source/user_guide/ug_implementation.rst [hunk 11]
```diff
@@ -122,6 +138,35 @@ and **global\_tx1.kmt** contain the :math:`\left<1^\circ\right>` POP
...
+   +----------------+-------+-------+-------+-------+
+   | area           | tarea | uarea | narea | earea |
+   +----------------+-------+-------+-------+-------+
+   | mask (logical) | tmask | umask | nmask | emask |
+   +----------------+-------+-------+-------+-------+
+   | mask (real)    |  hm   | uvm   | npm   | epm   |
+   +----------------+-------+-------+-------+-------+
+
+
 In CESM, the sea ice model may exchange coupling fluxes using a
```
Comment (eclare108213) [inline doc/source/user_guide/ug_implementation.rst:170]: Will we use this namelist variable 'gridcpl_file' for distinguishing between coupling on the C- and CD-grids?

**Review by eclare108213 — COMMENTED** [review]
Just comments for now.  I'm looking at this pretty closely because these changes are fundamental to the new capability, and I want to make sure that we're all on the same page re the assumptions made.  This looks great overall -- following responses to my questions/comments, we can probably merge it with some project-board reminders to check/clean up a few things later.

File: doc/source/user_guide/ug_implementation.rst (hunk 11 above)
Comment (apcraig) [reply doc/source/user_guide/ug_implementation.rst:170]: gridcpl_file is something totally different.  We have a new namelist, grid_system that can take on values of B, C, CD, etc as we move forward.

File: doc/source/user_guide/ug_case_settings.rst (hunk 10 above)
Comment (apcraig) [reply doc/source/user_guide/ug_case_settings.rst:243]: I think we need to completely separate the solver and the coupling.  The solver may have B, C, CD options, whatever.  The coupling will have it's own set of options, maybe fields come in on the T grid, U grid, N grid, E grid, and that could be quite a mix and match for different coupled systems, different forcing files, etc.  It could even be that fields come in on a grid that is somehow not associated with the B, CD, etc grids.  We want a lot of flexibility for the coupling fields.

File: cicecore/cicedynB/infrastructure/ice_grid.F90 (hunk 9 above)
Comment (apcraig) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:1981]: The only different between the to_u and t2ugrid_vector methods was a halo update.  So now the halo updated is required to be done outside the averager and the 4 old averages are all gone, but reimplemented in a bfb way in grid_average_X2Y (we can change the name if folks don't like that one).

File: cicecore/cicedynB/infrastructure/ice_grid.F90 (hunk 8 above)
Comment (apcraig) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:1953]: I agree and think you're right.  Just frustrating that the trig calcs are not bit-for-bit if called in different orders.  I don't understand that one!

File: cicecore/cicedynB/infrastructure/ice_grid.F90 (hunk 7 above)
Comment (apcraig) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:70]: CICE only reads HTN and HTE.  All other grid lengths are derived from those.  The current extension to N and E grids is done the "same way" as how the U dx,dy values are computed.  That method is far from ideal (does not conserve area locally), but it's what we've had for a long time.  I think it's worthwhile looking into whether we can improve the solution at some point down the road.

File: cicecore/cicedynB/dynamics/ice_dyn_shared.F90 (hunk 6 above)
Comment (apcraig) [reply cicecore/cicedynB/dynamics/ice_dyn_shared.F90:166]: Correct.  I figured that was the easiest thing to do for now.  And I added uveln and vvele as the "extra" arrays for the CD version.  This may not be what we want, but it's a starting point.

File: cicecore/cicedynB/analysis/ice_history.F90 (hunk 5 above)
Comment (apcraig) [reply cicecore/cicedynB/analysis/ice_history.F90]: This is one of the things we said we needed to review.

File: cicecore/cicedynB/analysis/ice_history.F90 (hunk 4 above)
Comment (apcraig) [reply cicecore/cicedynB/analysis/ice_history.F90]: Correct.  The U grid is now defined at the corner point.  The u and v grid could be the U, N, or E grid.  We do need to work on making this clearer in history output, but we do write the history coordinates as part of the output as well, so that helps.

File: cicecore/cicedynB/analysis/ice_history_shared.F90 (hunk 2 above)
Comment (apcraig) [reply cicecore/cicedynB/analysis/ice_history_shared.F90:169]: I've added a tick box to the project page.

File: doc/source/user_guide/ug_implementation.rst (hunk 11 above)
Comment (apcraig) [reply doc/source/user_guide/ug_implementation.rst:170]: I answered too quickly.  gridcpl_file allows a model to couple on a completely different grid.  In RASM we use this to coupled on an extended grid of the internal grid and we setup ice fraction to 0 at the extended grid points.  This helps the regional coupled model significantly.  The location of the forcing/coupling fields is totally different.  We'll need to create namelist to handle those options.  None of that is done yet, we still need to make some decisions about how we want to do that, and it's not critical for the initial implementation, but does need to be done at some point (it could even be after we have the CD implementation on the trunk).  For now, coupling/forcing is always assumed to be on the T grid.  Extending that to support other grids for subsets of forcing/coupling fields should not be difficult.

**Comment by dabail10** [comment]
This is great. I am in an all day meeting M-W this week, but I will try to take a look. I was also going to work on some history stuff this week. Looks like you've got a lot of done already.

File: cicecore/cicedynB/analysis/ice_history.F90 (hunk 4 above)
Comment (eclare108213) [reply cicecore/cicedynB/analysis/ice_history.F90]: I'd like this one fixed for this PR, please.  It's confusing (to me) otherwise.  BTW did you spot check the history output from a run to make sure it's working as expected, including existing variables like uvel?

File: cicecore/cicedynB/infrastructure/ice_grid.F90 (hunk 7 above)
Comment (eclare108213) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:70]: I'll admit I'm being a little picky here, but here's what's available from the grid file:

> [eclare@ba-fe1 gx3]$ ncdump -h grid_gx3.nc
> netcdf grid_gx3 {
> dimensions:
>         y = 116 ;
>         x = 100 ;
> variables:
>         double ulat(y, x) ;
>                 ulat:units = "rad" ;
>         double ulon(y, x) ;
>                 ulon:units = "rad" ;
>         double htn(y, x) ;
>                 htn:units = "cm" ;
>         double hte(y, x) ;
>                 hte:units = "cm" ;
>         double hus(y, x) ;
>                 hus:units = "cm" ;
>         double huw(y, x) ;
>                 huw:units = "cm" ;
>         double angle(y, x) ;
>                 angle:units = "rad" ;

It would be prudent to read in hus and huw, and then set dxe = hus and dyn = huw, at least for now.  How different are the resulting grid lengths, from using the averaging method?  Hopefully not very much...  but does that help the area conservation issue?  Using existing grid file variables would save some pretty substantial chunks of code in ice_grid.F90, for calculating dxe and dyn.

File: cicecore/cicedynB/infrastructure/ice_grid.F90 (hunk 9 above)
Comment (eclare108213) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:1981]: The grid_average_X2Y name is fine -- it's better than the previous subroutine name.  My main point here is that I don't think we're doing vector averages correctly in old or new code.

File: doc/source/user_guide/ug_case_settings.rst (hunk 10 above)
Comment (eclare108213) [reply doc/source/user_guide/ug_case_settings.rst:243]: Yes, agreed.  @JFLemieux73 can correct me on this, but I'm not expecting to have a purely C-grid solver.

File: cicecore/cicedynB/analysis/ice_history.F90 (hunk 4 above)
Comment (apcraig) [reply cicecore/cicedynB/analysis/ice_history.F90]: OK, I'll refactor it a bit.  I have checked that I can vary the underlying coordinate system for history fields based on grid_subsystem.  One other thing, these comments that are created for variables are not actually written to the history file right now.  I'm not sure why, they certainly could be.  But it's not implemented in the netcdf or pio IO versions.

Comment (eclare108213) [reply cicecore/cicedynB/analysis/ice_history.F90]: Weird that they aren't written -- they definitely should be!  Another thing to follow up on...

File: cicecore/cicedynB/infrastructure/ice_grid.F90 (hunk 7 above)
Comment (apcraig) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:70]: We're opening a huge can of worms here.  Using HUS and HUW will almost certainly improve the accuracy of the grid and local area conservation.  But, if we read them, then dxt and dyt should also probably change.  Yes, dxe=HUS.  But then dxt should be 0.5*(HUSij + HUSi-1j) and that will change all answers, make CICE inconsistent with POP (and impact POP/CICE coupling conservation?), and possibly break some usage if users have grid files that don't include HUS and HUW.  I'm willing to go there, but we should also carefully examine the solutions that are produced.  We could run some qc as a starting point.  Do we want to do this now or is that a separate task/issue to tackle separately, either before or after CD grid?  And why weren't dxt and dyt based on HUS and HUW in prior implementations?

Comment (eclare108213) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:70]: Okay.  We are identifying lots of issues to look into later.  So POP doesn't use HUS and HUW either?  This goes WAY back, farther than my memory serves, but for the infrastructure stuff I copied POP as closely as possible, and there was an effort back then to reduce memory usage wherever we could, so maybe we chose to not read them in.  dxt and dyt are calculated from HTN and HTE (e.g. dxt=0.5*(HTNij+HTNij-1)).  dxu and dyu would be averaged similarly from HUS and HUW.  I'm fine with continuing to do the analogous calculations for now, recognizing that there are some sanity checks we could do using the originally calculated grid lengths from the files.

Comment (apcraig) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:70]: I looked at POP.  It computes the dx, dy just like CICE currently.  But it also reads HUS, HUW and stores it internally and uses it for a few computations not directly related to grid lengths.  Very odd.  I propose we add a namelist variable that allows us to switch back and forth between "old" and "new" grid computations.

**Comment by dabail10** [comment]
Most groups are moving away from POP. So, I don't think we should necessarily retain the POP-style infrastructure. Just my $0.02.

**Comment by apcraig** [comment]
I largely agree @dabail10.  But for those that still do couple with POP, we probably need to keep the backwards compatibility.  RASM is certainly in this situation and likely to stay there for a least a little while longer.  RASM updates CICE regularly, and continues to use older-ish versions of POP.  It could be 5-10 years before all users of those applications (including probably recent versions of CESM) have fully moved on.  It could also be that we can upgrade the CICE grid calcs without any problem to POP coupling, but that's something the Consortium may not be able to determine universally very easily.

**Comment by dabail10** [comment]
The coupler checks TLAT, TLON and the mask. So, as long as these are consistent I believe we can do whatever we want.

**Comment by apcraig** [comment]
Could be, but answers will change in CICE and you'll end up with the same TLON, TLAT, and tmask in POP and CICE, but tarea will not be the same in both models.  For coupling systems that do direct coupling of fluxes with no area corrections, this will lead to a problem as the system will no longer conserve between CICE and POP.  Maybe there aren't any/many of those applications anymore, but again, we don't really know.

**Comment by dabail10** [comment]
I guess I was thinking TAREA would be the same as well. Nonetheless, your point is taken. I do like the idea of adding a flag to maintain backward compatibility.

**Comment by eclare108213** [comment]
> Most groups are moving away from POP. So, I don't think we should necessarily retain the POP-style infrastructure. Just my $0.02.

I totally agree.  POP frequently comes up in answer to questions like "why is so and so that way?"  We are definitely no longer tied to POP, but that said, there were really good reasons why things were done they way they were in POP.  I'm not sure any ocean model is as computationally efficient, now, as POP for what it could (and still does) do.  But ocean modeling has gotten a lot more complex (and sea ice modeling too).

File: cicecore/cicedynB/analysis/ice_history.F90 (hunk 4 above)
Comment (apcraig) [reply cicecore/cicedynB/analysis/ice_history.F90]: I have updated the history writing and comments are now being written.

**Review by eclare108213 — APPROVED** [review]

**Comment by phil-blain** [comment]
I   forgot earlier that tomorrow is holiday; I won't have time to take a look at this before Friday.

**Review by eclare108213 — APPROVED** [review]

**Comment by apcraig** [comment]
I'll merge now.  @phil-blain, feel free to review and provide feedback, I'm sure we'll be creating several PRs over the next week or two.  Thanks.

File: configuration/scripts/tests/unittest_suite.ts [hunk 12]
```diff
@@ -6,3 +6,5 @@ unittest       gx3     1x1x25x29x16 sumchk
 unittest       tx1     8x1          sumchk
 unittest       gx3     4x1          bcstchk
 unittest       gx3     1x1          bcstchk
+unittest       gx3     8x2          gridavgchk,dwblockall
+unittest       gx3     8x2          gridavgchk
```
Comment (phil-blain) [inline configuration/scripts/tests/unittest_suite.ts]: General comment about this 'unittest' test suite: We should probably run it as part of the CI, if possible. And maybe, instruct developers/contributors to run it as part of the standard testing procedure (i.e. in addition to the base_suite ) ?

**Review by phil-blain — COMMENTED** [review]
I had a look. Thanks for all this work, it's a major first step.
